### PR TITLE
asymcute: fix one byte out-of-bounds access in _len_get

### PR DESCRIPTION
**Pull request by nmeum**
### Contribution description

As per Section 5.2.1 of the MQTT-SN specification, the MQTT-SN length header is either 1- or 3-octet long. If it is 3-octet long then the first octet is `0x01`. The asymcute implementation currently only checks that the incoming packet is at least 2-octet long before attempting to parse it (`MIN_PKT_LE`N):

https://github.com/RIOT-OS/RIOT/blob/4dcb8edcc8f985ba25f83cccf0a7c8d2ceee74f2/sys/net/application_layer/asymcute/asymcute.c#L578

However, if the first octet is `0x01` the packet must be more than 3 octet long in order to be valid. Since asymcute does not check this it reads one octet beyond the packet data (via `byteorder_bebuftohs` which reads a `uint16_t` in `_len_get`) for a 2-octet packet where the first octet has the value `0x01`. This commit fixes this issue by adding an additional sanity check to `_len_get`.

### Testing procedure

I think this should (hopefully) be obvious from reading the code and comparing consulting the aforementioned section in the MQTT-SN specification.

Also note that emcute has an explicit check for this:

https://github.com/RIOT-OS/RIOT/blob/b0f4781e1547357c5fbe5256f940c95b324a711c/sys/net/application_layer/emcute/emcute.c#L531-L534

### Issues/PRs references

None.

### CODE REVIEW

File: sys/net/application_layer/asymcute/asymcute.c [hunk 1]
```diff
@@ -83,13 +83,15 @@ static size_t _len_set(uint8_t *buf, size_t len)
...
 
-static size_t _len_get(uint8_t *buf, size_t *len)
+static ssize_t _len_get(uint8_t *buf, size_t pkt_len, size_t *len)
 {
     if (buf[0] != 0x01) {
         *len = (uint16_t)buf[0];
         return 1;
     }
     else {
+        if (pkt_len <= 3)
```
Comment (miri64) [inline sys/net/application_layer/asymcute/asymcute.c:93]: Why `<=` and not `<`? From what I gather of the code below, it tries to read a 2 byte buffer from position 1 of the packet, so `pkt_len == 3` should be legal and for anything packet validity related, there is [a check in `_on_pkt()`](https://github.com/RIOT-OS/RIOT/pull/18433/files#diff-d63d8a165e6b370754ebb39bfeca42f1ed38f7c8a53a59fae1cb4db14fc7d8e6R590)

Comment (nmeum) [reply sys/net/application_layer/asymcute/asymcute.c]: My thinking was just that the message needs data apart from the length header, but you are right. This is caught by the check in `_on_pkt()` and hence `<` should be sufficient. Thanks for pointing this out, I adjusted the check accordingly.

**Review by miri64 — APPROVED** [review]
Looks good and change makes sense.

**Comment by miri64** [comment]
:exclamation: This may cause a semantic merge conflict with https://github.com/RIOT-OS/RIOT/pull/18434, once that is merged: `_get_len()` gains another parameter, so only trust Murdock if it has successfully build both with either of one in master (and check if the master Murdock used includes the other PR).

**Comment by miri64** [comment]
Build error is unrelated.

**Comment by benpicco** [comment]
This will now need a rebase anyway

**Comment by nmeum** [comment]
> This will now need a rebase anyway

Rebased and made the necessary adjustments for the new `_len_get` function prototype.

**Comment by nmeum** [comment]
> This will now need a rebase anyway

Rebased and made the necessary adjustments for the `_len_get` function prototype.

File: sys/net/application_layer/asymcute/asymcute.c [hunk 2]
```diff
@@ -107,7 +110,7 @@ static uint16_t _msg_id_next(asymcute_con_t *con)
 static uint8_t _req_type(asymcute_req_t *req)
 {
     size_t len;
-    size_t pos = _len_get(req->data, &len);
+    size_t pos = _len_get(req->data, req->data_len, &len);
     return req->data[pos];
```
Comment (nmeum) [inline sys/net/application_layer/asymcute/asymcute.c:114]: Might make sense to add an assert here that pos is != 1, pos should have been verified in _on_pkt anyhow but just to be sure...

Comment (miri64) [reply sys/net/application_layer/asymcute/asymcute.c]: Could you also add a comment to the code why the assertion must be true, please?

Comment (nmeum) [reply sys/net/application_layer/asymcute/asymcute.c]: Done.

Comment (miri64) [reply sys/net/application_layer/asymcute/asymcute.c]: Thanks!

**Comment by miri64** [comment]
This needs a backport to the 2022.07 branch as well.

**Pull request merged**